frame/staking/src/offchain_election.rs

```diff
@@ -73,42 +88,112 @@ pub(crate) const DEFAULT_LONGEVITY: u64 = 25;
-					Err("recently executed.")
-				}
-				Some(Some(head)) if now > head + threshold => {
-					// we can run again now. Write the new head.
-					Ok(now)
-				}
-				_ => {
-					// value doesn't exists. Probably this node just booted up. Write, and run
-					Ok(now)
-				}
+
+	let mutate_stat = storage.mutate(|maybe_head: Option<Option<T::BlockNumber>>| {
+		match maybe_head {
+			Some(Some(head)) if now < head => Err(OffchainElectionError::Fork),
+			Some(Some(head)) if now >= head && now <= head + threshold => {
+				Err(OffchainElectionError::TooRecent)
+			}
+			Some(Some(head)) if now > head + threshold => {
+				// we can allow again now. Write the new head.
+				Ok(now)
+			}
+			_ => {
+				// value doesn't exists. Probably this node just booted up. Write, and allow.
+				Ok(now)
 			}
-		});
+		}
+	});
 
+	crate::log!(trace, "attempting to acquire the OCW lock at {:?} = {:?}", now, mutate_stat);
 	match mutate_stat {
 		// all good
 		Ok(Ok(_)) => Ok(()),
 		// failed to write.
-		Ok(Err(_)) => Err("failed to write to offchain db."),
+		Ok(Err(_)) => Err(OffchainElectionError::DBWriteFailed),
 		// fork etc.
 		Err(why) => Err(why),
 	}
 }
 
+/// Storage path for the solution `call`.
+pub(crate) const OFFCHAIN_QUEUED_CALL: &[u8] = b"parity/staking-election/call";
+
+/// Save a given call OCW storage.
+pub(crate) fn save_solution<T: Config>(call: Call<T>) -> Result<(), OffchainElectionError> {
+	let storage = StorageValueRef::persistent(&OFFCHAIN_QUEUED_CALL);
+	let set_outcome = storage.mutate::<_, OffchainElectionError, _>(|maybe_call| {
+		match maybe_call {
+			// under any case, write the new value. This means that we don't need to clear the
+			// solution. Each era, it will be overwritten.
+			_ => Ok(call),
+		}
+	});
+
+	match set_outcome {
+		Ok(Ok(_)) => Ok(()),
+		// failed to write.
+		Ok(Err(_)) => Err(OffchainElectionError::DBWriteFailed),
+		_ => {
+			// Defensive only: should not happen. Inner mutate closure always returns ok.
+			Err(OffchainElectionError::Unreachable)
+		}
+	}
+}
+
+/// Get a saved OCW solution, if it exists.
+pub(crate) fn get_solution<T: Config>() -> Option<Call<T>> {
+	StorageValueRef::persistent(&OFFCHAIN_QUEUED_CALL).get().flatten()
+}
+
+/// Submit a given solution as an unsigned transaction.
+pub(crate) fn submit_solution<T: Config>(call: Call<T>) -> Result<(), OffchainElectionError> {
+	SubmitTransaction::<T, Call<T>>::submit_unsigned_transaction(call.into())
+		.map_err(|_| OffchainElectionError::PoolSubmissionFailed)
+}
+
+/// Ensure that the given solution call belongs to the current era. Returns `Ok(call)` if so to be
+/// used with `Result::and`.
+pub(crate) fn ensure_solution_is_recent<T: Config>(
+	call: Call<T>,
+) -> Result<Call<T>, OffchainElectionError> {
+	let current_era = <Module<T>>::current_era().unwrap_or_default();
+	match call {
+		Call::submit_election_solution_unsigned(_, _, _, era, _) if era == current_era => Ok(call),
+		_ => Err(OffchainElectionError::SolutionOld),
```

Comment:
It might also be for the next era I suppose and in such case the error might be a bit misleading. IT should be rare though (might happen if we decode garbage from the storage) and I don't think it's critical, but you could simply add both eras there to the error type so that the user can have a bit more details.